CIP-????/README.md

```diff
@@ -0,0 +1,476 @@
+
+A prototype implementation already exists and is functional.
+
+## Appendices
+
+### Appendix 1: Changes to Typed Plutus Core
+
+While Typed Plutus Core is not part of the specification of Cardano, it is still interesting and informative to give the changes here.
+Plutus Core was originally conceived of as a typed language, and making sure that we can express the changes cleanly in a typed setting means that we ensure that the semantics make sense and that it will continue to be easy to compile to Plutus Core from typed languages.
+
+We add the following new type constructors to the type language of Typed Plutus Core:
+```
+(sum tys)
+(prod tys)
+```
+These correspond to sum and product types.
+While we are not using sums-and-products at the term level, we can keep it at the type level since that does not affect evaluation.
+
+We add the following new term constructors to the Typed Plutus Core language:
+```
+(constr ty i ts)
+(case ty t cs)
+```
+
+These are identical to their untyped cousins, except that they include a type annotation for the type of the whole term. 
+These make the typing rules much simpler, as otherwise we lack enough information to pin down the whole type.
+
+The typing rules for the new terms are:
+
+$$
+\frac{\Gamma \vdash t_i : ty_i,\ rty = (\mathrm{sum\ s_1 \ldots s_e \ldots s_m}), s_e = (\mathrm{prod}\ ty_1 \ldots ty_n)}{\Gamma \vdash (\mathrm{constr}\ rty\ e\ t_1 \ldots t_n) : rty}
+$$
+
+$$
+\frac{\Gamma \vdash t : tty,\ tty = (\mathrm{sum}\ s_1 \ldots s_n), s_i = (\mathrm{prod}\ p_{i_1} \ldots p_{i_m}),\ \Gamma \vdash c_i : p_{i_1} \rightarrow \ldots \rightarrow p_{i_m} \rightarrow rty}{\Gamma \vdash (\mathrm{case}\ rty\ t\ c_1 \ldots c_n) : rty}
+$$
+
+The reduction rules are essentially the same since the types do not affect evaluation.
+
+### Appendix 2: Performance Analysis
+
+Performance testing of the various prototypes revealed the following interesting facts:
+1. Scott encoding is surprisingly performant.
+2. It's tricky to say why sums-of-products is so much better.
+3. Binding case variables makes a modest difference, but we can achieve the same in other ways.
+
+Let's look at these in turn.
+
+#### Why is Scott encoding so good?
+
+To see why Scott encoding is so good, let's look at what happens when we 1) construct, and 2) destruct a typical datatype value, using sums-of-products but _without_ binding case variables.
+We'll do this in the typed setting for clarity.
+
+Let’s consider four programs, corresponding to creation and destruction of `data XYZ = MkXY X Y | MkZ Z`.
+
+**P1: Scott-encoded construction**
+
+```
+[
+  -- Scott-encoded constructor for MkXY
+  (\(x : X) (y : Y) . (/\ R . \(xyc : X -> Y -> R) (zc : Z -> R) . xyc x y) 
+  xx 
+  yy
+]
+```
+
+**P2: Scott-encoded destruction**
+
+```
+[ 
+  { P1 r} 
+  -- case alternatives
+  (\(x : X) (y : Y) -> b1) 
+  (\(z : Z) -> b2
+]
+```
+
+**P3: Explicit construction**
+
+```
+constr 1 xx yy
```

Comment:
... and here it seems like indexing starts from 1.

And the evaluation and typing rules all start from 1.

Could you explicitly specify whether indexing starts from 0 or 1, and if it's 1 explain why you've chosen the wrong option :)